expression/builtin_other.go

```diff
@@ -106,11 +142,49 @@ func (c *inFunctionClass) getFunction(ctx sessionctx.Context, args []Expression)
 // builtinInIntSig see https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_in
 type builtinInIntSig struct {
 	baseBuiltinFunc
+	nonConstArgs []Expression
+	hashSet      map[int64]bool
+	threshold    int
+}
+
+func (b *builtinInIntSig) buildHashMapForConstArgs(ctx sessionctx.Context) error {
+	b.nonConstArgs = make([]Expression, 0, len(b.args))
+	b.nonConstArgs = append(b.nonConstArgs, b.args[0])
+	b.hashSet = make(map[int64]bool, len(b.args)-1)
+	count := 0
+	for i := 1; i < len(b.args); i++ {
+		if b.args[i].ConstItem() {
+			val, isNull, err := b.args[i].EvalInt(ctx, chunk.Row{})
+			if err != nil {
+				return err
+			}
+			if isNull {
+				b.nonConstArgs = append(b.nonConstArgs, b.args[i])
+				continue
+			}
```

Comment:
It seems that you can save a a simple flag in b to indicate that "whether here is a NULL in constant args".

Here is the implementation in tikv:

https://github.com/tikv/tikv/pull/6000/files#diff-1c2ffa236b010ac4d3e49ce35fbb12a7R127

Reply:
The behavior of NULL value in `in`:

<img width="314" alt="image" src="https://user-images.githubusercontent.com/9161438/71008376-75136a00-2123-11ea-8ddd-d914a426e5b6.png">